corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,283 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+            self.user.update_metadata(uncategorized_data)
+
+        # Clear blank user data so that it can be purged by remove_unused_custom_fields_from_users_task
+        for key in dict(data, **uncategorized_data):
+            value = self.user.metadata[key]
+            if value is None or value == '':
+                self.user.pop_metadata(key)
+
+    def update_language(self, language):
+        self.user.language = language
+        self.logger.add_changes({'language': language})
+
+    def update_email(self, email):
+        self.user.email = email.lower()
+        self.logger.add_changes({'email': self.user.email})
+
+    def update_status(self, is_active):
+        self.user.is_active = is_active
+        self.logger.add_changes({'is_active': is_active})
+
+    def update_locations(self, location_codes, domain_info):
+        from corehq.apps.user_importer.importer import (
+            check_modified_user_loc,
+            find_location_id,
+            get_location_from_site_code
+        )
+
+        location_ids = find_location_id(location_codes, domain_info.location_cache)
+        users_current_primary_location_id = self.user.location_id
+        locations_updated, primary_loc_removed = check_modified_user_loc(location_ids,
+                                                                         self.user.location_id,
+                                                                         self.user.assigned_location_ids)
+        if primary_loc_removed:
+            self.user.unset_location(commit=False)
+        if locations_updated:
+            self.user.reset_locations(location_ids, commit=False)
+            self.logger.add_changes({'assigned_location_ids': location_ids})
+            if location_ids:
+                location_names = [get_location_from_site_code(code, domain_info.location_cache).name
+                                  for code in location_codes]
+                self.logger.add_info(_(f"Assigned locations: {location_names}"))
+        # log this after assigned locations are updated, which can re-set primary location
+        if self.user.location_id != users_current_primary_location_id:
+            self.logger.add_changes({'location_id': self.user.location_id})
+            if self.user.location_id:
+                users_updated_primary_location_name = get_user_primary_location_name(self.user, self.user_domain)
+                self.logger.add_info(
+                    _(f"Primary location: {users_updated_primary_location_name}"))
+
+    def update_role(self, role, domain_info):
+        role_qualified_id = domain_info.roles_by_name[role]
+        user_current_role = self.user.get_role(domain=self.user_domain)
+        self.role_updated = not (user_current_role
+                                 and user_current_role.get_qualified_id() == role_qualified_id)
+        if self.role_updated:
+            self.user.set_role(self.user_domain, role_qualified_id)
+
+    def save(self):
```

Comment:
nit: `save` doesn't seem like the right name here. I think this could be `save_log` as well.

Reply:
https://github.com/dimagi/commcare-hq/pull/29935/commits/c7e526f7f23045a00126aa4456157fbf80ac45c9